File: vocabcraft-mcp/src/vocabcraft_mcp/tools/review.py
# src/vocabcraft_mcp/tools/review.py
"""复习排程 Tool - 基于遗忘曲线（SM-2）

根据词汇的复习状态查询到期词汇并生成复习计划。
评分时由 quiz.grade_quiz 调用 algorithms.compute_next_review 更新记忆状态，
再通过 crud.update_vocab 回写 review_state。

到期定义: review_state.next_review <= 今天（YYYY-MM-DD 字符串比较）。
"""

from vocabcraft_mcp.algorithms import _now_utc
from vocabcraft_mcp.tools.crud import get_storage


def schedule_review(vocab_id: str = "", language: str = "") -> dict:
    """生成复习计划

    - 指定 vocab_id：返回该词汇的复习状态与到期日
    - 未指定 vocab_id：返回所有到期词汇列表，按到期日升序

    Args:
        vocab_id: 指定词汇 ID，空串则查询全部到期词
        language: 按语种过滤（en/zh 等），空串则不过滤

    Returns:
        指定 vocab_id 时: {vocab_id, word, review_state, due_date, is_due}
        未指定时: {today, due_count, due_words: [{vocab_id, word, due_date}]}
        词汇不存在时: {error}
    """
    storage = get_storage()
    today = _now_utc().date().isoformat()

    # 单词汇模式：返回复习状态
    if vocab_id:
        v = storage.load_vocab(vocab_id)
        if v is None:
            return {"error": f"词汇不存在: {vocab_id}"}
        due_date = v.review_state.next_review
        return {
            "vocab_id": v.id,
            "word": v.structured.word,
            "review_state": v.review_state.model_dump(),
            "due_date": due_date,
            "is_due": bool(due_date) and due_date <= today,
        }

    # 全局模式：查询所有到期词汇
    all_vocabs = storage.get_all_vocabs_for_statistics()
    due = [
        v for v in all_vocabs
        if v.review_state.next_review and v.review_state.next_review <= today
        and (not language or v.structured.language == language)
    ]
    # 按到期日升序，早到期的优先
    due.sort(key=lambda v: v.review_state.next_review)

    return {
        "today": today,
        "due_count": len(due),
        "due_words": [
            {
                "vocab_id": v.id,
                "word": v.structured.word,
                "due_date": v.review_state.next_review,
            }
            for v in due
        ],
    }

File: vocabcraft.plugin/vocabcraft-mcp-go/internal/tools/review.go
package tools

import (
	"time"

	"github.com/yecllsl/vocabcraft-mcp-go/internal/store"
)

type Review struct {
	store *store.Store
}

func NewReview(s *store.Store) *Review {
	return &Review{store: s}
}

func (r *Review) ScheduleReview(vocabID, language string) map[string]any {
	today := time.Now().UTC().Format("2006-01-02")

	if vocabID != "" {
		v, err := r.store.LoadVocab(vocabID)
		if err != nil {
			return map[string]any{"error": "词汇不存在: " + vocabID}
		}
		return map[string]any{
			"vocab_id": v.ID,
			"word":     v.Structured.Word,
			"review_state": map[string]any{
				"ease_factor":     v.ReviewState.EaseFactor,
				"interval":        v.ReviewState.Interval,
				"repetitions":     v.ReviewState.Repetitions,
				"next_review":     v.ReviewState.NextReview,
				"last_review":     v.ReviewState.LastReview,
				"last_word_grade": v.ReviewState.LastWordGrade,
			},
			"due_date": v.ReviewState.NextReview,
			"is_due":   v.ReviewState.NextReview != "" && v.ReviewState.NextReview <= today,
		}
	}

	vocabs, err := r.store.AllVocabs()
	if err != nil {
		return map[string]any{"error": err.Error()}
	}

	type dueItem struct {
		VocabID string `json:"vocab_id"`
		Word    string `json:"word"`
		DueDate string `json:"due_date"`
	}
	var due []dueItem
	for _, v := range vocabs {
		if v.ReviewState.NextReview == "" || v.ReviewState.NextReview > today {
			continue
		}
		if language != "" && v.Structured.Language != language {
			continue
		}
		due = append(due, dueItem{
			VocabID: v.ID,
			Word:    v.Structured.Word,
			DueDate: v.ReviewState.NextReview,
		})
	}

	// Sort by due_date, then vocab_id
	for i := 1; i < len(due); i++ {
		for j := i; j > 0; j-- {
			if due[j].DueDate < due[j-1].DueDate ||
				(due[j].DueDate == due[j-1].DueDate && due[j].VocabID < due[j-1].VocabID) {
				due[j], due[j-1] = due[j-1], due[j]
			}
		}
	}

	return map[string]any{
		"today":     today,
		"due_count": len(due),
		"due_words": due,
	}
}
